File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/json/JSONModelLoader.ts
///<reference path="../Callback.ts"/>
///<reference path="../KObject.ts"/>
///<reference path="../KView.ts"/>
///<reference path="../ModelLoader.ts"/>
///<reference path="../abs/AbstractKObject.ts"/>
///<reference path="../data/AccessMode.ts"/>
///<reference path="../meta/MetaAttribute.ts"/>
///<reference path="../meta/MetaReference.ts"/>
///<reference path="../meta/MetaType.ts"/>
///<reference path="../time/TimeTree.ts"/>
///<reference path="../../../../../java/util/JUArrayList.ts"/>
///<reference path="../../../../../java/util/JUHashMap.ts"/>
///<reference path="../../../../../java/util/JUHashSet.ts"/>
///<reference path="../../../../../java/util/JUList.ts"/>
///<reference path="../../../../../java/util/JUMap.ts"/>
///<reference path="../../../../../java/util/JUSet.ts"/>

class JSONModelLoader implements ModelLoader {

  private _factory: KView = null;

  constructor(p_factory: KView) {
    this._factory = p_factory;
  }

  public static load(payload: string, factory: KView, callback: Callback<KObject<any,any>>): KObject<any,any> {
    var lexer: Lexer = new Lexer(payload);
    var loaded: KObject<any,any>[] = new Array();
    JSONModelLoader.loadObjects(lexer, factory, {on:function(objs: JUList<KObject<any,any>>){
    loaded[0] = objs.get(0);
}});
    return loaded[0];
  }

  private static loadObjects(lexer: Lexer, factory: KView, callback: Callback<JUList<KObject<any,any>>>): void {
    var loaded: JUList<KObject<any,any>> = new JUArrayList<KObject<any,any>>();
    var alls: JUList<JUMap<string, any>> = new JUArrayList<JUMap<string, any>>();
    var content: JUMap<string, any> = new JUHashMap<string, any>();
    var currentAttributeName: string = null;
    var arrayPayload: JUSet<string> = null;
    var currentToken: JsonToken = lexer.nextToken();
    while (currentToken.tokenType() != Type.EOF){
      if (currentToken.tokenType().equals(Type.LEFT_BRACKET)) {
        arrayPayload = new JUHashSet<string>();
      } else {
        if (currentToken.tokenType().equals(Type.RIGHT_BRACKET)) {
          content.put(currentAttributeName, arrayPayload);
          arrayPayload = null;
          currentAttributeName = null;
        } else {
          if (currentToken.tokenType().equals(Type.LEFT_BRACE)) {
            content = new JUHashMap<string, any>();
          } else {
            if (currentToken.tokenType().equals(Type.RIGHT_BRACE)) {
              alls.add(content);
              content = new JUHashMap<string, any>();
            } else {
              if (currentToken.tokenType().equals(Type.VALUE)) {
                if (currentAttributeName == null) {
                  currentAttributeName = currentToken.value().toString();
                } else {
                  if (arrayPayload == null) {
                    content.put(currentAttributeName, currentToken.value().toString());
                    currentAttributeName = null;
                  } else {
                    arrayPayload.add(currentToken.value().toString());
                  }
                }
              }
            }
          }
        }
      }
      currentToken = lexer.nextToken();
    }
    var keys: number[] = new Array();
    for (var i: number = 0; i < keys.length; i++) {
      var kid: number = Long.parseLong(alls.get(i).get(JSONModelSerializer.KEY_UUID).toString());
      keys[i] = kid;
    }
    factory.dimension().timeTrees(keys, {on:function(timeTrees: TimeTree[]){
    for (var i: number = 0; i < alls.size(); i++) {
      var elem: JUMap<string, any> = alls.get(i);
      var meta: string = elem.get(JSONModelSerializer.KEY_META).toString();
      var kid: number = Long.parseLong(elem.get(JSONModelSerializer.KEY_UUID).toString());
      var isRoot: boolean = false;
      var root: any = elem.get(JSONModelSerializer.KEY_ROOT);
      if (root != null) {
        isRoot = root.toString().equals("true");
      }
      var timeTree: TimeTree = timeTrees[i];
      timeTree.insert(factory.now());
      var current: KObject<any,any> = factory.createProxy(factory.metaClass(meta), timeTree, kid);
      if (isRoot) {
        (<AbstractKObject<any,any>>current).setRoot(true);
      }
      loaded.add(current);
      var payloadObj: any[] = factory.dimension().universe().storage().raw(current, AccessMode.WRITE);
      //TODO resolve for-each cycle
      var k: string;
      for (k in elem.keySet()) {
        var att: MetaAttribute = current.metaAttribute(k);
        if (att != null) {
          payloadObj[att.index()] = JSONModelLoader.convertRaw(att, elem.get(k));
        } else {
          var ref: MetaReference = current.metaReference(k);
          if (ref != null) {
            if (ref.single()) {
              var refPayloadSingle: number;
              try {
                refPayloadSingle = Long.parseLong(elem.get(k).toString());
                payloadObj[ref.index()] = refPayloadSingle;
              } catch ($ex$) {
                if ($ex$ instanceof Exception) {
                  var e: Exception = <Exception>$ex$;
                  e.printStackTrace();
                }
               }
            } else {
              try {
                var plainRawList: JUSet<string> = <JUSet<string>>elem.get(k);
                var convertedRaw: JUSet<number> = new JUHashSet<number>();
                //TODO resolve for-each cycle
                var plainRaw: string;
                for (plainRaw in plainRawList) {
                  try {
                    var converted: number = Long.parseLong(plainRaw);
                    convertedRaw.add(converted);
                  } catch ($ex$) {
                    if ($ex$ instanceof Exception) {
                      var e: Exception = <Exception>$ex$;
                      e.printStackTrace();
                    }
                   }
                }
                payloadObj[ref.index()] = convertedRaw;
              } catch ($ex$) {
                if ($ex$ instanceof Exception) {
                  var e: Exception = <Exception>$ex$;
                  e.printStackTrace();
                }
               }
            }
          }
        }
      }
    }
    if (callback != null) {
      callback.on(loaded);
    }
}});
  }

  public load(payload: string, callback: Callback<Throwable>): void {
    if (payload == null) {
      callback.on(null);
    } else {
      var lexer: Lexer = new Lexer(payload);
      var currentToken: JsonToken = lexer.nextToken();
      if (currentToken.tokenType() != Type.LEFT_BRACKET) {
        callback.on(null);
      } else {
        JSONModelLoader.loadObjects(lexer, this._factory, {on:function(kObjects: JUList<KObject<any,any>>){
        callback.on(null);
}});
      }
    }
  }

  public static convertRaw(attribute: MetaAttribute, raw: any): any {
    try {
      if (attribute.metaType().equals(MetaType.STRING)) {
        return raw.toString();
      } else {
        if (attribute.metaType().equals(MetaType.LONG)) {
          return Long.parseLong(raw.toString());
        } else {
          if (attribute.metaType().equals(MetaType.INT)) {
            return Integer.parseInt(raw.toString());
          } else {
            if (attribute.metaType().equals(MetaType.BOOL)) {
              return raw.toString().equals("true");
            } else {
              if (attribute.metaType().equals(MetaType.SHORT)) {
                return Short.parseShort(raw.toString());
              } else {
                if (attribute.metaType().equals(MetaType.DOUBLE)) {
                  return Double.parseDouble(raw.toString());
                } else {
                  if (attribute.metaType().equals(MetaType.FLOAT)) {
                    return Float.parseFloat(raw.toString());
                  } else {
                    return null;
                  }
                }
              }
            }
          }
        }
      }
    } catch ($ex$) {
      if ($ex$ instanceof Exception) {
        var e: Exception = <Exception>$ex$;
        e.printStackTrace();
        return null;
      }
     }
  }

}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/json/JsonModelLoader.java
package org.kevoree.modeling.api.json;

import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.KConfig;
import org.kevoree.modeling.api.KObject;
import org.kevoree.modeling.api.KView;
import org.kevoree.modeling.api.abs.AbstractKView;
import org.kevoree.modeling.api.abs.AbstractMetaReference;
import org.kevoree.modeling.api.data.cache.KCacheEntry;
import org.kevoree.modeling.api.data.manager.AccessMode;
import org.kevoree.modeling.api.data.manager.Index;
import org.kevoree.modeling.api.data.manager.JsonRaw;
import org.kevoree.modeling.api.map.LongLongHashMap;
import org.kevoree.modeling.api.map.StringHashMap;
import org.kevoree.modeling.api.map.StringHashMapCallBack;
import org.kevoree.modeling.api.meta.Meta;
import org.kevoree.modeling.api.meta.MetaAttribute;
import org.kevoree.modeling.api.meta.MetaClass;
import org.kevoree.modeling.api.meta.MetaModel;
import org.kevoree.modeling.api.meta.MetaType;

import java.util.ArrayList;
import java.util.List;

public class JsonModelLoader {

    /**
     * @native ts
     * if (payload == null) {
     * callback(null);
     * } else {
     * var metaModel: org.kevoree.modeling.api.meta.MetaModel = factory.universe().model().metaModel();
     * var toLoadObj = JSON.parse(payload);
     * var rootElem = [];
     * var mappedKeys: org.kevoree.modeling.api.map.LongLongHashMap = new org.kevoree.modeling.api.map.LongLongHashMap(toLoadObj.length, org.kevoree.modeling.api.KConfig.CACHE_LOAD_FACTOR);
     * for(var i = 0; i < toLoadObj.length; i++) {
     * var elem = toLoadObj[i];
     * var kid = elem[org.kevoree.modeling.api.json.JsonModelSerializer.KEY_UUID];
     * mappedKeys.put(<number>kid, factory.universe().model().manager().nextObjectKey());
     * }
     * for(var i = 0; i < toLoadObj.length; i++) {
     * var elemRaw = toLoadObj[i];
     * var elem2 = new org.kevoree.modeling.api.map.StringHashMap<any>(Object.keys(elemRaw).length, org.kevoree.modeling.api.KConfig.CACHE_LOAD_FACTOR);
     * for(var ik in elemRaw){ elem2[ik] = elemRaw[ik]; }
     * try {
     * org.kevoree.modeling.api.json.JsonModelLoader.loadObj(elem2, metaModel, factory, mappedKeys, rootElem);
     * } catch(e){ console.error(e); }
     * }
     * if (rootElem[0] != null) { factory.setRoot(rootElem[0]).then( (throwable : java.lang.Throwable) => { if (callback != null) { callback(throwable); }}); } else { if (callback != null) { callback(null); } }
     * }
     */
    public static void load(KView factory, String payload, final Callback<Throwable> callback) {
        if (payload == null) {
            callback.on(null);
        } else {
            MetaModel metaModel = factory.universe().model().metaModel();
            Lexer lexer = new Lexer(payload);
            JsonType currentToken = lexer.nextToken();
            if (currentToken != JsonType.LEFT_BRACKET) {
                callback.on(null);
            } else {
                final List<StringHashMap<Object>> alls = new ArrayList<StringHashMap<Object>>();
                StringHashMap<Object> content = new StringHashMap<Object>(KConfig.CACHE_INIT_SIZE, KConfig.CACHE_LOAD_FACTOR);
                String currentAttributeName = null;
                ArrayList<String> arrayPayload = null;
                currentToken = lexer.nextToken();
                while (currentToken != JsonType.EOF) {
                    if (currentToken.equals(JsonType.LEFT_BRACKET)) {
                        arrayPayload = new ArrayList<String>();
                    } else if (currentToken.equals(JsonType.RIGHT_BRACKET)) {
                        content.put(currentAttributeName, arrayPayload);
                        arrayPayload = null;
                        currentAttributeName = null;
                    } else if (currentToken.equals(JsonType.LEFT_BRACE)) {
                        content = new StringHashMap<Object>(KConfig.CACHE_INIT_SIZE, KConfig.CACHE_LOAD_FACTOR);
                    } else if (currentToken.equals(JsonType.RIGHT_BRACE)) {
                        alls.add(content);
                        content = new StringHashMap<Object>(KConfig.CACHE_INIT_SIZE, KConfig.CACHE_LOAD_FACTOR);
                    } else if (currentToken.equals(JsonType.VALUE)) {
                        if (currentAttributeName == null) {
                            currentAttributeName = lexer.lastValue();
                        } else {
                            if (arrayPayload == null) {
                                content.put(currentAttributeName, lexer.lastValue());
                                currentAttributeName = null;
                            } else {
                                arrayPayload.add(lexer.lastValue());
                            }
                        }
                    }
                    currentToken = lexer.nextToken();
                }
                final KObject[] rootElem = {null};
                LongLongHashMap mappedKeys = new LongLongHashMap(alls.size(), KConfig.CACHE_LOAD_FACTOR);
                for (int i = 0; i < alls.size(); i++) {
                    try {
                        StringHashMap<Object> elem = alls.get(i);
                        long kid = Long.parseLong(elem.get(JsonModelSerializer.KEY_UUID).toString());
                        mappedKeys.put(kid, factory.universe().model().manager().nextObjectKey());
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                for (int i = 0; i < alls.size(); i++) {
                    try {
                        StringHashMap<Object> elem = alls.get(i);
                        loadObj(elem, metaModel, factory, mappedKeys, rootElem);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                if (rootElem[0] != null) {
                    factory.setRoot(rootElem[0]).then(new Callback<Throwable>() {
                        @Override
                        public void on(Throwable throwable) {
                            if (callback != null) {
                                callback.on(throwable);
                            }
                        }
                    });
                } else {
                    if (callback != null) {
                        callback.on(null);
                    }
                }
            }
        }
    }

    private static void loadObj(StringHashMap<Object> p_param, MetaModel p_metaModel, KView p_factory, LongLongHashMap p_mappedKeys, KObject[] p_rootElem) {
        long kid = Long.parseLong(p_param.get(JsonModelSerializer.KEY_UUID).toString());
        String meta = p_param.get(JsonModelSerializer.KEY_META).toString();
        MetaClass metaClass = p_metaModel.metaClass(meta);
        KObject current = ((AbstractKView) p_factory).createProxy(metaClass, p_mappedKeys.get(kid));
        p_factory.universe().model().manager().initKObject(current, p_factory);
        KCacheEntry raw = p_factory.universe().model().manager().entry(current, AccessMode.WRITE);
        p_param.each(new StringHashMapCallBack<Object>() {
            @Override
            public void on(String metaKey, Object payload_content) {
                if (metaKey.equals(JsonModelSerializer.INBOUNDS_META)) {
                    try {
                        raw.set(Index.INBOUNDS_INDEX, transposeArr((ArrayList<String>) payload_content, p_mappedKeys));
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                } else if (metaKey.equals(JsonModelSerializer.PARENT_META)) {
                    try {
                        ArrayList<String> parentKeys = (ArrayList<String>) payload_content;
                        long[] parentTransposed = transposeArr(parentKeys,p_mappedKeys);
                        if(parentTransposed != null && parentTransposed.length >0 && parentTransposed[0]!=KConfig.NULL_LONG){
                            long[] parentKey = new long[1];
                            parentKey[0] = parentTransposed[0];
                            raw.set(Index.PARENT_INDEX, parentKey);
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                } else if (metaKey.equals(JsonModelSerializer.PARENT_REF_META)) {
                    try {
                        String parentRef_payload = payload_content.toString();
                        String[] elems = parentRef_payload.split(JsonRaw.SEP);
                        if (elems.length == 2) {
                            MetaClass foundMeta = p_metaModel.metaClass(elems[0].trim());
                            if (foundMeta != null) {
                                Meta metaReference = foundMeta.metaByName(elems[1].trim());
                                if (metaReference != null && metaReference instanceof AbstractMetaReference) {
                                    raw.set(Index.REF_IN_PARENT_INDEX, metaReference);
                                }
                            }
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                } else if (metaKey.equals(JsonModelSerializer.KEY_ROOT)) {
                    if ("true".equals(payload_content)) {
                        p_rootElem[0] = current;
                    }
                } else {
                    Meta metaElement = metaClass.metaByName(metaKey);
                    if (payload_content != null) {
                        if (metaElement != null && metaElement.metaType().equals(MetaType.ATTRIBUTE)) {
                            raw.set(metaElement.index(), ((MetaAttribute) metaElement).strategy().load(payload_content.toString(), (MetaAttribute) metaElement, p_factory.now()));
                        } else if (metaElement != null && metaElement instanceof AbstractMetaReference) {
                            try {
                                raw.set(metaElement.index(), transposeArr((ArrayList<String>) payload_content, p_mappedKeys));
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                    }
                }
            }
        });
    }

    /**
     * @native ts
     * if (plainRawSet == null) { return null; }
     * var convertedRaw: number[] = new Array();
     * for (var l in plainRawSet) {
     * try {
     * var converted: number = java.lang.Long.parseLong(plainRawSet[l]);
     * if (p_mappedKeys.containsKey(converted)) { converted = p_mappedKeys.get(converted); }
     * convertedRaw[l] = converted;
     * } catch ($ex$) {
     * if ($ex$ instanceof java.lang.Exception) {
     * var e: java.lang.Exception = <java.lang.Exception>$ex$;
     * e.printStackTrace();
     * }
     * }
     * }
     * return convertedRaw;
     */
    private static long[] transposeArr(ArrayList<String> plainRawSet, LongLongHashMap p_mappedKeys) {
        if (plainRawSet == null) {
            return null;
        }
        long[] convertedRaw = new long[plainRawSet.size()];
        for (int l = 0; l < plainRawSet.size(); l++) {
            try {
                long converted = Long.parseLong(plainRawSet.get(l));
                if (p_mappedKeys.containsKey(converted)) {
                    converted = p_mappedKeys.get(converted);
                }
                convertedRaw[l] = converted;
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return convertedRaw;
    }

}

